downstairs/src/main.rs

```diff
@@ -790,15 +761,28 @@ async fn proc(
                                     another_upstairs_active_tx.clone()
                                 );
                             }
-
+                            negotiated = 2;
                             fw.send(Message::YouAreNowActive(uuid)).await?;
                         }
                     }
+                    Some(Message::RegionInfoPlease) => {
+                        if negotiated != 2 {
+                            bail!("Received RegionInfo out of order {}",
+                                negotiated);
+                        }
+                        negotiated = 3;
+                        let rd = {
+                            let ds = ads.lock().await;
+                            ds.region.def()
+                        };
+                        fw.send(Message::RegionInfo(rd)).await?;
+                    }
                     Some(Message::LastFlush(last_flush)) => {
-                        // TODO: Make a proper negotiation connect flow.
-                        if !negotiated {
-                            bail!("expected HereIAm first");
+                        if negotiated != 3 {
+                            bail!("Received LastFlush out of order {}",
+                                negotiated);
                         }
+                        negotiated = 5;
```

Comment:
4?

Reply:
I'll add a comment to explain it, but it's because..
Well, wait. 
I think you are correct here, I could just go to 4 and since there are no more commands
coming after that, stop the loop on 4.

The final negotiation message should either be LastFlush or ExtentVersionsPlease, so, yeah.
I'll update that before I merge.